pkg/controllermanager/controller/credentialsbinding/reconciler.go

```diff
@@ -0,0 +1,258 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package credentialsbinding
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	securityv1alpha1 "github.com/gardener/gardener/pkg/apis/security/v1alpha1"
+	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
+	"github.com/gardener/gardener/pkg/controllerutils"
+)
+
+// Reconciler reconciles CredentialsBinding.
+type Reconciler struct {
+	Client   client.Client
+	Config   config.CredentialsBindingControllerConfiguration
+	Recorder record.EventRecorder
+}
+
+// Reconcile reconciles CredentialsBinding.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	ctx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	credentialsBinding := &securityv1alpha1.CredentialsBinding{}
+	if err := r.Client.Get(ctx, request.NamespacedName, credentialsBinding); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	// The deletionTimestamp labels a CredentialsBinding as intended to get deleted. Before deletion,
+	// it has to be ensured that no Shoots are depending on the CredentialsBinding anymore.
+	// When this happens the controller will remove the finalizers from the CredentialsBinding so that it can be garbage collected.
+	if credentialsBinding.DeletionTimestamp != nil {
```

Comment:
it would be nice to split the function here into `delete()` and `reconcile()` funcs